src/App.tsx

```diff
@@ -1,26 +1,224 @@
+
+  const addNewTodo = () => {
+    const titleTrimmed = title.trim();
+
+    if (!titleTrimmed) {
+      setError(Errors.EmptyTitle);
+
+      return;
+    }
+
+    const temp = {
+      title: titleTrimmed,
+      id: 0,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    setTempTodo(temp);
+    setError(null);
+
+    postTodo(temp)
+      .then(res => {
+        setTodos(prev => [...prev, res]);
+        setTitle('');
+      })
+      .catch(() => setError(Errors.Add))
+      .finally(() => {
+        setTempTodo(null);
+      });
+  };
+
+  const deleteCurrentTodo = useCallback((id: number) => {
+    setDeleteTodoId(id);
+    deleteTodo(id)
+      .then(() => {
+        setTodos(prev => prev.filter(todo => todo.id !== id));
+      })
+      .catch(() => {
+        setError(Errors.Delete);
+      })
+      .finally(() => setDeleteTodoId(null));
+  }, []);
+
+  const clearCompleted = async () => {
+    const todoIdsToClear: number[] = [];
+
+    try {
+      for (const todo of todos) {
+        if (todo.completed) {
+          try {
+            await deleteTodo(todo.id);
+            todoIdsToClear.push(todo.id);
+          } catch {
+            setError(Errors.Delete);
+          }
+        }
+      }
+
+      setTodos(prevState =>
+        prevState.filter(todo => !todoIdsToClear.includes(todo.id)),
+      );
+    } catch {
+      setError(Errors.Delete);
+    }
+  };
+
+  const toggleAll = async () => {
+    try {
+      const isAnyUncompleted = todos.some(todo => !todo.completed);
+
+      const completedValue = isAnyUncompleted ? true : false;
+
+      await Promise.all(
+        todos.map(async todo => {
+          try {
+            if (todo.completed !== completedValue) {
+              await setCompletedTodo({
+                id: todo.id,
+                completed: completedValue,
+              });
+            }
+          } catch {
+            setError(Errors.Update);
+          }
+        }),
+      );
+
+      setTodos(prevTodos =>
+        prevTodos.map(prevTodo => ({
+          ...prevTodo,
+          completed: completedValue,
+        })),
+      );
+    } catch {
+      setError(Errors.Update);
+    }
+  };
+
+  const toggleCompleted = async (id: number) => {
+    try {
+      await setCompletedTodo({
+        id,
+        completed: !todos.find(todo => todo.id === id)?.completed,
+      });
+
+      setTodos(prevTodos =>
+        prevTodos.map(todo =>
+          todo.id === id ? { ...todo, completed: !todo.completed } : todo,
+        ),
+      );
+    } catch {
+      setError(Errors.Update);
+    }
+  };
+
+  const updateTodoTitle = (todoId: number, newTitle: string) => {
+    const updatedTodos = todos.map(todo => {
+      if (todo.id === todoId) {
+        return { ...todo, title: newTitle };
+      }
+
+      return todo;
+    });
+
+    setTodos(updatedTodos);
+  };
+
+  useEffect(() => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [tempTodo, todos]);
 
```

Comment:
General comment on a code structure: it's always better to have a decent order for your hooks - helpers. For example, it can be

```
useRefs
useStates
useEffects
helpers
```

or 

```
useRefs
useStates
helpers
useEffects
```

Try not to mix declarations like:

```
useRef
useState
helper
useEffect
helpers
useEffect
```